File: scripts/sandbox/lume-clone.js
'use strict';

const fs = require('fs');
const os = require('os');
const path = require('path');
const crypto = require('crypto');
const { spawnSync } = require('child_process');

const MAX_AUX_BYTES = 16 * 1024 ** 2;
const MAX_DISK_BYTES = 8 * 1024 ** 4;
const SOURCE = path.join(__dirname, 'lume-clone.swift');
const PINNED_LUME_VERSION = '0.5.1';
const BLOCKED = new Set(['sessions.json', '.provisioning', 'resize.lock.json']);
const hash = bytes => crypto.createHash('sha256').update(bytes).digest('hex');

function assertPrivateDirectory(directory) {
  const info = fs.lstatSync(directory);
  if (!info.isDirectory() || info.isSymbolicLink() || info.uid !== process.getuid() || (info.mode & 0o077)) {
    throw new Error('Untrusted helper cache');
  }
}

function invoke(executable, args, timeout) {
  const result = spawnSync(executable, args, {
    encoding: 'utf8', timeout, killSignal: 'SIGKILL', maxBuffer: 65536, shell: false,
    env: { PATH: '/usr/bin:/bin:/usr/sbin:/sbin', HOME: os.homedir(), LC_ALL: 'C', LANG: 'C' },
  });
  if (result.error || result.signal || result.status !== 0) throw new Error('Helper operation failed');
  return result.stdout;
}

function helperIdentity(binary) {
  const info = fs.lstatSync(binary);
  if (!info.isFile() || info.isSymbolicLink() || info.uid !== process.getuid() || (info.mode & 0o077) || info.size > 16 * 1024 ** 2) {
    throw new Error('Untrusted helper binary');
  }
  return hash(fs.readFileSync(binary));
}

function compileHelper() {
  const source = fs.readFileSync(SOURCE);
  const digest = hash(Buffer.concat([source, Buffer.from(`\n${process.arch}\n${PINNED_LUME_VERSION}`)]));
  const cache = `/private/tmp/ecc-lume-clone-${process.getuid()}-${digest}`;
  try { fs.mkdirSync(cache, { mode: 0o700 }); } catch (error) { if (error.code !== 'EEXIST') throw error; }
  assertPrivateDirectory(cache);
  const binary = path.join(cache, 'helper');
  const record = path.join(cache, 'artifact.json');
  if (fs.existsSync(binary) && fs.existsSync(record)) {
    const metadata = JSON.parse(fs.readFileSync(record, 'utf8'));
    if (metadata.source_digest !== digest || metadata.binary_digest !== helperIdentity(binary)) throw new Error('Helper cache integrity failure');
    return Object.freeze({ binary, digest: metadata.binary_digest });
  }
  const staging = fs.mkdtempSync(path.join(cache, 'build-'));
  try {
    const snapshot = path.join(staging, 'source.swift');
    fs.writeFileSync(snapshot, source, { mode: 0o600, flag: 'wx' });
    const output = path.join(staging, 'helper');
    invoke('/usr/bin/xcrun', ['swiftc', '-O', '-framework', 'Virtualization', snapshot, '-o', output], 60_000);
    fs.chmodSync(output, 0o700);
    const binaryDigest = helperIdentity(output);
    fs.renameSync(output, binary);
    fs.writeFileSync(record, JSON.stringify({ source_digest: digest, binary_digest: binaryDigest }), { mode: 0o600 });
    return Object.freeze({ binary, digest: binaryDigest });
  } finally {
    fs.rmSync(staging, { recursive: true, force: true });
  }
}

function existingDirectory(directory) {
  if (typeof directory !== 'string' || !path.isAbsolute(directory) || directory.includes('\0') || fs.realpathSync(directory) !== directory) {
    throw new Error('Invalid or symlinked directory');
  }
  const info = fs.lstatSync(directory);
  if (!info.isDirectory() || info.isSymbolicLink()) throw new Error('Invalid directory');
  return info;
}

function validatePaths(seedPath, destPath) {
  const source = existingDirectory(seedPath);
  if (typeof destPath !== 'string' || !path.isAbsolute(destPath) || destPath.includes('\0') ||
      !/^ecc-(sandbox|explore|fabric)-lume-[a-zA-Z0-9-]+$/.test(path.basename(destPath)) || path.normalize(destPath) !== destPath || fs.existsSync(destPath)) {
    throw new Error('Invalid or occupied destination');
  }
  const parent = existingDirectory(path.dirname(destPath));
  if (source.dev !== parent.dev) throw new Error('Cross-device clone unsupported');
  const entries = fs.readdirSync(seedPath);
  if (entries.length > 64 || !['disk.img', 'nvram.bin', 'config.json'].every(name => entries.includes(name))) throw new Error('Invalid seed contents');
  let auxiliaryBytes = 0;
  for (const name of entries) {
    if (BLOCKED.has(name)) throw new Error('Source state is unsafe to clone');
    const file = fs.lstatSync(path.join(seedPath, name));
    if (!file.isFile() || file.isSymbolicLink() || !Number.isSafeInteger(file.size) || file.size < 0 ||
        file.size > (name === 'disk.img' ? MAX_DISK_BYTES : name === 'nvram.bin' ? 64 * 1024 ** 2 : MAX_AUX_BYTES)) throw new Error('Unsupported seed entry');
    if (name !== 'disk.img') auxiliaryBytes += file.size;
  }
  if (auxiliaryBytes > 80 * 1024 ** 2) throw new Error('Oversized seed metadata');
  return Object.freeze({ sourceDevice: source.dev, sourceInode: source.ino, parentDevice: parent.dev, parentInode: parent.ino });
}

function verifyCompatibility(options) {
  // Match the lifecycle runner's PATH lookup, without invoking a shell.
  const requested = options.lumeExecutable || 'lume';
  const candidates = path.isAbsolute(requested) ? [requested] : requested === 'lume'
    ? (process.env.PATH || '').split(path.delimiter).map(directory => path.resolve(directory || '.', requested)) : [];
  const executable = candidates.find(candidate => {
    try { fs.accessSync(candidate, fs.constants.X_OK); return fs.statSync(candidate).isFile(); } catch { return false; }
  });
  if (typeof executable !== 'string' || !path.isAbsolute(executable) || !fs.statSync(executable).isFile()) throw new Error('Lume executable unavailable');
  const run = options.run || spawnSync;
  const settings = { encoding: 'utf8', shell: false, timeout: 2000, maxBuffer: 65536 };
  const version = run(executable, ['--version'], settings);
  if (version.error || version.signal || version.status !== 0 || typeof version.stdout !== 'string' || version.stdout.trim() !== PINNED_LUME_VERSION) throw new Error('Lume version unsupported');
  const pressure = run('/usr/sbin/sysctl', ['-n', 'kern.memorystatus_vm_pressure_level'], settings);
  if (pressure.error || pressure.signal || pressure.status !== 0 || typeof pressure.stdout !== 'string' || pressure.stdout.trim() !== '1') throw new Error('Compilation refused under unknown or elevated memory pressure');
}

function checkedStopped(callback) {
  if (typeof callback !== 'function' || callback() !== true) throw new Error('Source stopped state unverified');
}

function denied(code, message, ownership = {}) {
  return Object.freeze({ ok: false, code, message, copy_method: 'clonefile-required', owned_destination: false, cleanup_pass: true, ...ownership });
}
function fail(receipt) {
  const error = new Error(receipt.message);
  error.receipt = receipt;
  throw error;
}
function executeClone(helper, seedPath, destPath, identity) {
  const expected = [identity.sourceDevice, identity.sourceInode, identity.parentDevice, identity.parentInode].map(String);
  const result = spawnSync(helper.binary, ['clone', seedPath, destPath, ...expected], {
    encoding: 'utf8', timeout: 30_000, killSignal: 'SIGKILL', maxBuffer: 65536, shell: false,
    env: { PATH: '/usr/bin:/bin:/usr/sbin:/sbin', HOME: os.homedir(), LC_ALL: 'C', LANG: 'C' },
  });
  return validateCloneReceipt(result);
}
function validateCloneReceipt(result) {
  let receipt;
  try { receipt = JSON.parse(result.stdout); } catch { receipt = null; }
  if (result.error || result.signal || !receipt || typeof receipt.owned_destination !== 'boolean' ||
      !(receipt.cleanup_pass === null || typeof receipt.cleanup_pass === 'boolean') || receipt.copy_method !== 'clonefile-required') {
    fail(denied('lume_cow_unverified', 'Clone helper was interrupted or returned incomplete evidence. Inspect the destination; ownership could not be verified.', { cleanup_pass: false }));
  }
  if (result.status !== 0 || receipt.ok !== true) {
    fail(denied('lume_cow_failed', 'Forced copy-on-write cloning failed without regular-copy fallback.', { owned_destination: receipt.owned_destination, cleanup_pass: receipt.cleanup_pass }));
  }
  if (receipt.owned_destination !== true || receipt.cleanup_pass !== null) {
    fail(denied('lume_cow_unverified', 'Clone helper returned inconsistent ownership evidence. Inspect the destination before retrying.', { cleanup_pass: false }));
  }
  return Object.freeze({ ok: true, code: 'lume_cow_cloned', message: 'VM files cloned with fresh machine and network identities; full-copy fallback was disabled.', copy_method: 'clonefile-required', owned_destination: true, cleanup_pass: null });
}

// ECC-owned compatibility optimization for the pinned Lume version. Unsupported
// preparation keeps the caller's full-copy disk budget. A prepared operation
// never falls back to copying and never starts a VM.
function prepareLumeClone(seedPath, destPath, options = {}) {
  const unavailable = (message, reason = 'unsupported_platform') => Object.freeze({
    supported: false, cloneBytes: null, code: 'lume_cow_unavailable', reason, message,
    clone: () => fail(denied('lume_cow_unavailable', message)),
  });
  if ((options.platform || process.platform) !== 'darwin') return unavailable('Forced copy-on-write cloning requires macOS. Keep the full-copy disk budget.');
  let identity;
  let helper;
  let phase = 'source_state_unverified';
  try {
    checkedStopped(options.verifySourceStopped);
    phase = 'seed_layout_invalid';
    identity = validatePaths(seedPath, destPath);
    phase = 'lume_compatibility_unverified';
    verifyCompatibility(options);
    phase = 'helper_build_failed';
    helper = compileHelper();
    phase = 'cow_probe_failed';
    const proof = JSON.parse(invoke(helper.binary, ['probe', path.dirname(destPath)], 5_000));
    if (proof.ok !== true || proof.copy_method !== 'clonefile-required') throw new Error('Clone capability unproven');
  } catch {
    return unavailable(`Forced copy-on-write cloning could not be verified (${phase}). Keep the full-copy disk budget; check the stopped seed, storage, Lume version, memory pressure, and Xcode command-line tools.`, phase);
  }
  let consumed = false;
  return Object.freeze({
    supported: true, cloneBytes: 0, code: 'lume_cow_ready',
    message: 'Per-file copy-on-write cloning is verified on the destination volume; full-copy fallback is disabled.',
    clone() {
      if (consumed) fail(denied('lume_cow_consumed', 'This prepared clone operation has already been consumed.'));
      consumed = true;
      try {
        checkedStopped(options.verifySourceStopped);
        const current = validatePaths(seedPath, destPath);
        if (JSON.stringify(current) !== JSON.stringify(identity) || helperIdentity(helper.binary) !== helper.digest) throw new Error('Clone inputs changed');
        return executeClone(helper, seedPath, destPath, identity);
      } catch (error) {
        if (error.receipt) throw error;
        fail(denied('lume_cow_failed', 'Forced copy-on-write cloning failed validation before destination creation. No regular-copy fallback was attempted.'));
      }
    },
  });
}

module.exports = { prepareLumeClone, validateCloneReceipt };

File: scripts/sandbox/lume-clone.swift
// ECC-owned Lume 0.5.1 clone compatibility helper. Never boots a guest.
// Pinned upstream: trycua/cua@7e13c6437acb1e2193548d36f520b3519ceee05c
// Per-file clonefile is mandatory. There is deliberately no copy fallback.
import Foundation
import Darwin
import Virtualization

struct CloneFailure: Error { let message: String }
var ownedDestination = false
var cleanupPass: Bool? = true
func require(_ condition: Bool, _ message: String) throws {
    if !condition { throw CloneFailure(message: message) }
}
func info(_ descriptor: Int32) throws -> stat {
    var value = stat()
    try require(fstat(descriptor, &value) == 0, "Cannot inspect opened file")
    return value
}
func openDirectory(_ path: String) throws -> Int32 {
    guard let canonical = realpath(path, nil) else { throw CloneFailure(message: "Cannot resolve directory") }
    defer { free(canonical) }
    try require(String(cString: canonical) == path, "Symlinked directory refused")
    let fd = open(path, O_RDONLY | O_DIRECTORY | O_NOFOLLOW | O_CLOEXEC)
    try require(fd >= 0, "Cannot open directory")
    return fd
}
func regular(_ value: stat) -> Bool { (value.st_mode & S_IFMT) == S_IFREG }
func acquireGuard(_ parent: Int32, _ name: String) throws -> Int32 {
    let fd = openat(parent, ".\(name).resize.guard", O_RDWR | O_CREAT | O_NOFOLLOW | O_CLOEXEC, 0o600)
    try require(fd >= 0, "Cannot open VM lifecycle guard")
    do {
        let value = try info(fd)
        try require(regular(value) && value.st_uid == getuid() && value.st_nlink == 1, "Unsafe VM lifecycle guard")
        try require(flock(fd, LOCK_EX | LOCK_NB) == 0, "VM lifecycle guard is busy")
        return fd
    } catch { close(fd); throw error }
}
func releaseGuard(_ fd: Int32) { _ = flock(fd, LOCK_UN); close(fd) }
func entryNames(_ path: String) throws -> [String] {
    let names = try FileManager.default.contentsOfDirectory(atPath: path).sorted()
    try require(names.count <= 64, "Too many seed entries")
    try require(["disk.img", "nvram.bin", "config.json"].allSatisfy { names.contains($0) }, "Seed files missing")
    let blocked: Set<String> = ["sessions.json", ".provisioning", "resize.lock.json"]
    try require(names.allSatisfy { !blocked.contains($0) && !$0.contains("/") }, "Seed session, provisioning, or resize state present")
    return names
}
func openedFiles(_ directory: Int32, _ names: [String]) throws -> [(String, Int32)] {
    var opened: [(String, Int32)] = []
    do {
        var auxiliary: Int64 = 0
        for name in names {
            let fd = openat(directory, name, O_RDONLY | O_NOFOLLOW | O_NONBLOCK | O_CLOEXEC)
            try require(fd >= 0, "Cannot open seed entry")
            opened.append((name, fd))
            let value = try info(fd)
            let limit: Int64 = name == "disk.img" ? 8 * 1024 * 1024 * 1024 * 1024 : (name == "nvram.bin" ? 64 * 1024 * 1024 : 16 * 1024 * 1024)
            try require(regular(value) && value.st_size >= 0 && value.st_size <= limit, "Seed entry is special or oversized")
            if name != "disk.img" { auxiliary += value.st_size }
        }
        try require(auxiliary <= 80 * 1024 * 1024, "Seed metadata too large")
        return opened
    } catch { for (_, fd) in opened { close(fd) }; throw error }
}
func writeAll(_ fd: Int32, _ bytes: Data) throws {
    try bytes.withUnsafeBytes { raw in
        var count = 0
        while count < bytes.count {
            let result = write(fd, raw.baseAddress!.advanced(by: count), bytes.count - count)
            if result < 0 && errno == EINTR { continue }
            try require(result > 0, "Cannot write identity metadata")
            count += result
        }
    }
}
func freshConfig(_ directory: Int32) throws {
    let fd = openat(directory, "config.json", O_RDONLY | O_NOFOLLOW | O_CLOEXEC)
    try require(fd >= 0, "Missing cloned configuration")
    defer { close(fd) }
    let value = try info(fd)
    try require(value.st_size > 0 && value.st_size <= 1024 * 1024, "Invalid configuration size")
    var data = Data(count: Int(value.st_size))
    let count = data.withUnsafeMutableBytes { read(fd, $0.baseAddress, $0.count) }
    try require(count == data.count, "Cannot read complete configuration")
    guard var config = try JSONSerialization.jsonObject(with: data) as? [String: Any],
          let os = config["os"] as? String, os.lowercased() == "macos",
          let hardware = config["hardwareModel"] as? String, Data(base64Encoded: hardware)?.isEmpty == false,
          let machine = config["machineIdentifier"] as? String, Data(base64Encoded: machine)?.isEmpty == false,
          config["macAddress"] is String else {
        throw CloneFailure(message: "Unsupported seed configuration")
    }
    // Match LumeController.clone: retain hardware model and disk/NVRAM pairing.
    config["macAddress"] = VZMACAddress.randomLocallyAdministered().string
    config["machineIdentifier"] = VZMacMachineIdentifier().dataRepresentation.base64EncodedString()
    let bytes = try JSONSerialization.data(withJSONObject: config, options: [.sortedKeys])
    let temporary = ".ecc-config-new"
    let output = openat(directory, temporary, O_WRONLY | O_CREAT | O_EXCL | O_NOFOLLOW | O_CLOEXEC, 0o600)
    try require(output >= 0, "Cannot create fresh configuration")
    defer { close(output); _ = unlinkat(directory, temporary, 0) }
    try writeAll(output, bytes)
    try require(fsync(output) == 0, "Cannot sync fresh configuration")
    try require(renameat(directory, temporary, directory, "config.json") == 0, "Cannot publish fresh configuration")
}
func cleanupOwned(_ parent: Int32, _ name: String, _ directory: Int32, _ identity: stat, _ entries: [String]) -> Bool {
    var current = stat()
    guard fstatat(parent, name, &current, AT_SYMLINK_NOFOLLOW) == 0,
          current.st_ino == identity.st_ino, current.st_dev == identity.st_dev else { return false }
    for entry in entries { _ = unlinkat(directory, entry, 0) }
    _ = unlinkat(directory, ".ecc-config-new", 0)
    return unlinkat(parent, name, AT_REMOVEDIR) == 0
}
func cloneVM(_ source: String, _ destination: String, _ expected: [String]) throws {
    let sourceURL = URL(fileURLWithPath: source)
    let destURL = URL(fileURLWithPath: destination)
    let destName = destURL.lastPathComponent
    try require(destName.range(of: "^ecc-(sandbox|explore|fabric)-lume-[a-zA-Z0-9-]+$", options: .regularExpression) != nil, "Destination is outside ECC namespace")
    let sourceParent = try openDirectory(sourceURL.deletingLastPathComponent().path)
    defer { close(sourceParent) }
    let destParent = try openDirectory(destURL.deletingLastPathComponent().path)
    defer { close(destParent) }
    let sourceGuard = try acquireGuard(sourceParent, sourceURL.lastPathComponent)
    defer { releaseGuard(sourceGuard) }
    let destGuard = try acquireGuard(destParent, destName)
    defer { releaseGuard(destGuard) }
    let sourceDir = try openDirectory(source)
    defer { close(sourceDir) }
    let sourceIdentity = try info(sourceDir)
    let parentIdentity = try info(destParent)
    try require([String(sourceIdentity.st_dev), String(sourceIdentity.st_ino), String(parentIdentity.st_dev), String(parentIdentity.st_ino)] == expected, "Prepared directory identities changed")
    try require(sourceIdentity.st_dev == parentIdentity.st_dev, "Cross-volume clone refused")
    let names = try entryNames(source)
    let files = try openedFiles(sourceDir, names)
    defer { for (_, fd) in files { close(fd) } }
    // Publish no destination until all file types, sizes, and lifecycle state pass.
    try require(mkdirat(destParent, destName, 0o700) == 0, "Destination already exists or cannot be created")
    ownedDestination = true
    cleanupPass = false
    let destinationDir = openat(destParent, destName, O_RDONLY | O_DIRECTORY | O_NOFOLLOW | O_CLOEXEC)
    if destinationDir < 0 { cleanupPass = unlinkat(destParent, destName, AT_REMOVEDIR) == 0; throw CloneFailure(message: "Cannot open owned destination") }
    defer { close(destinationDir) }
    let identity = try info(destinationDir)
    var created: [String] = []
    var success = false
    defer { if !success { cleanupPass = cleanupOwned(destParent, destName, destinationDir, identity, created) } }
    for (name, sourceFile) in files {
        try require(fclonefileat(sourceFile, destinationDir, name, UInt32(CLONE_NOFOLLOW_ANY)) == 0, "Required copy-on-write clone failed")
        created.append(name)
    }
    try freshConfig(destinationDir)
    try require(fsync(destinationDir) == 0, "Cannot sync cloned directory")
    success = true
    cleanupPass = nil
}
func probe(_ parentPath: String) throws {
    let parent = try openDirectory(parentPath)
    defer { close(parent) }
    let name = ".ecc-clone-probe-\(UUID().uuidString)"
    try require(mkdirat(parent, name, 0o700) == 0, "Cannot create clone probe")
    let directory = openat(parent, name, O_RDONLY | O_DIRECTORY | O_NOFOLLOW | O_CLOEXEC)
    if directory < 0 { _ = unlinkat(parent, name, AT_REMOVEDIR); throw CloneFailure(message: "Cannot open probe") }
    defer { close(directory); _ = unlinkat(parent, name, AT_REMOVEDIR) }
    defer { _ = unlinkat(directory, "original", 0); _ = unlinkat(directory, "cloned", 0) }
    let original = openat(directory, "original", O_RDWR | O_CREAT | O_EXCL | O_NOFOLLOW | O_CLOEXEC, 0o600)
    try require(original >= 0, "Cannot create original probe file")
    defer { close(original) }
    try writeAll(original, Data([65]))
    try require(fclonefileat(original, directory, "cloned", UInt32(CLONE_NOFOLLOW_ANY)) == 0, "Filesystem cannot guarantee copy-on-write")
    let cloned = openat(directory, "cloned", O_WRONLY | O_NOFOLLOW | O_CLOEXEC)
    try require(cloned >= 0, "Cannot open clone probe")
    defer { close(cloned) }
    try writeAll(cloned, Data([66]))
    var byte: UInt8 = 0
    try require(pread(original, &byte, 1, 0) == 1 && byte == 65, "Clone writes are not isolated")
}
do {
    let args = CommandLine.arguments
    if args.count == 3 && args[1] == "probe" { try probe(args[2]) }
    else if args.count == 8 && args[1] == "clone" { try cloneVM(args[2], args[3], Array(args[4...7])) }
    else { throw CloneFailure(message: "Invalid helper invocation") }
    let output = try JSONSerialization.data(withJSONObject: ["ok": true, "copy_method": "clonefile-required", "owned_destination": ownedDestination, "cleanup_pass": cleanupPass as Any? ?? NSNull()])
    print(String(data: output, encoding: .utf8)!)
} catch {
    // No paths, configuration values, or credentials in output.
    let reason = (error as? CloneFailure)?.message ?? "Invalid clone metadata"
    let output = try! JSONSerialization.data(withJSONObject: ["ok": false, "copy_method": "clonefile-required", "message": reason, "owned_destination": ownedDestination, "cleanup_pass": cleanupPass as Any? ?? NSNull()])
    print(String(data: output, encoding: .utf8)!)
    exit(1)
}
